### PR TITLE
Add pydantic models to GBNF grammar generator. For "Instructor" library like use of llama.cpp models.

### PR DESCRIPTION
**Pull request by Maximilian-Winter**
This adds a pydantic model to gbnf grammar generator. Can be used for easy function calling and other stuff. Generates a JSON grammar based on Pydantic BaseModel classes. Can also convert type hinted Python functions into pydantic models and this is into a grammar. I developed it for my llama-cpp-agent framework.

https://github.com/Maximilian-Winter/llama-cpp-agent

### CODE REVIEW

**Review by ggerganov — APPROVED** [review]
Thanks for the example! It would be useful if there are basic usage instructions. Also, if you can come-up with a basic unit test, we can add it to the CI and make sure the script

**Comment by Maximilian-Winter** [comment]
@ggerganov Should I simply add a new file with the examples? Or should I add them in the file?

**Comment by ggerganov** [comment]
A new file should be fine. If the examples are small enough, you can simply add them to the PR description or the Wiki and add a link in the script header. Anyway you think makes sense

**Comment by Maximilian-Winter** [comment]
@ggerganov I have added a function calling example with detailed usage instruction. I can also add a example for structured output in JSON form based on Pydantic models. Like a Knowledge Graph.

**Comment by Maximilian-Winter** [comment]
@ggerganov Added another example for simple structured output.

File: examples/pydantic_models_to_grammar.py
Comment (ggerganov) [inline examples/pydantic_models_to_grammar.py:1]: Change the name of the script to use dashes: `pydantic-models-to-grammar.py`

Comment (Maximilian-Winter) [reply examples/pydantic_models_to_grammar.py:1]: I fixed it very ugly with import lib.

Comment (ggerganov) [reply examples/pydantic_models_to_grammar.py:1]: Thanks. I realize the "dashes in filenames" convention seems to be incompatible with Python scripts - an oversight on my side. It's fine to keep the underscores for simplicity - we already have a few exceptions in this regard.

**Comment by Solido** [comment]
Hi, thank you for the feature @Maximilian-Winter.

I've confronted with this message while calling `python3 pydantic-models-to-grammar-examples.py`. What did I miss?

``` python

  File "llama.cpp/examples/pydantic-models-to-grammar-examples.py", line 74, in <module>
    gbnf_grammar, documentation = generate_gbnf_grammar_and_documentation(
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "llama.cpp/examples/pydantic_models_to_grammar.py", line 1078, in generate_gbnf_grammar_and_documentation
    documentation = generate_markdown_documentation(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "llama.cpp/examples/pydantic_models_to_grammar.py", line 719, in generate_markdown_documentation
    documentation += generate_field_markdown(
                     ^^^^^^^^^^^^^^^^^^^^^^^^
  File "llama.cpp/examples/pydantic_models_to_grammar.py", line 752, in generate_field_markdown
    field_info = model.model_fields.get(field_name)
                 ^^^^^^^^^^^^^^^^^^
AttributeError: type object 'SendMessageToUser' has no attribute 'model_fields'
```

**Comment by cebtenzzre** [comment]
> I've confronted with this message while calling `python3 pydantic-models-to-grammar-examples.py`. What did I miss?

You need to install at *least* (not sure the exact minimum) pydantic v2.0, which renamed `__fields__` to `model_fields`. You must currently have pydantic 1.x.

**Comment by Solido** [comment]
The non python dev that I'm thank you @cebtenzzre 
pydantic 2.5.3 did it.

**Pull request merged**